dependency/vault_pki.go

```diff
@@ -66,40 +64,68 @@ func (d *VaultPKIQuery) Fetch(clients *ClientSet, opts *QueryOptions,
 	default:
 	}
 
-	var encPEM []byte
-	var err error
-	if encPEM, err = os.ReadFile(d.filePath); err != nil || len(encPEM) == 0 {
-		// no need to write to file as it is the template dest
-		encPEM, err = d.fetchPEM(clients)
-	}
-	if err != nil {
-		return nil, nil, errors.Wrap(err, d.String())
-	}
+	needsRenewal := fmt.Errorf("needs renewal")
+	getPEM := func(renew bool) ([]byte, error) {
+		var encPEM []byte
+		var err error
+		encPEM, err = os.ReadFile(d.filePath)
+		if renew || err != nil || len(encPEM) == 0 {
+			encPEM, err = d.fetchPEM(clients)
+			// no need to write cert to file as it is the template dest
+		}
+		if err != nil {
+			return nil, err
+		}
 
-	block, cert, err := getCert(encPEM)
-	if err != nil {
-		return nil, nil, err
+		block, cert, err := getCert(encPEM)
+		if err != nil {
+			return nil, err
+		}
+
+		if sleepFor, ok := goodFor(cert); ok {
+			d.sleepCh <- sleepFor
+			return pem.EncodeToMemory(block), nil
+		}
+		return []byte{}, needsRenewal
 	}
 
-	if sleepFor, ok := goodFor(cert); ok {
-		d.sleepCh <- sleepFor
-		return respWithMetadata(string(pem.EncodeToMemory(block)))
+	pemBytes, err := getPEM(false)
+	switch err {
+	case nil:
+	case needsRenewal:
+		pemBytes, err = getPEM(true)
+		if err != nil {
+			return nil, nil, err
+		}
+	default:
+		return nil, nil, err
 	}
-	d.pemBlock = nil
-	return "", &ResponseMetadata{
-		LastContact: 0,
-		LastIndex:   0,
-	}, nil
+	return respWithMetadata(string(pemBytes))
 }
 
-// returns time left in 90% of the original lease
-// boolean returns false if time has already past
+// returns time left in ~90% of the original lease and a boolean
+// that returns false if cert needs renewing, true otherwise
 func goodFor(cert *x509.Certificate) (time.Duration, bool) {
 	start, end := cert.NotBefore.Unix(), cert.NotAfter.Unix()
-	// use 90% of the cert duration
-	goodfor := ((end - start) * 9) / 10
-	sleepFor := time.Until(time.Unix(start+goodfor, 0))
-	return sleepFor, sleepFor > 0
+	now := time.Now().UTC().Unix()
+	if end <= now || start >= now {
+		return 0, false // expired or not good yet
+	}
+	lifespan := end - start
+	duration := end - now          // duration remaining
+	gooddur := (duration * 9) / 10 // 90% of duration
+	mindur := (lifespan / 10)      // 10% of lifespan
+	if gooddur <= mindur {
+		return 0, false // almost expired, get a new one
+	}
+	if gooddur > 100 { // add jitter if big enough for it to matter
+		r := rand.New(rand.NewSource(time.Now().UnixNano()))
```

Comment:
I don't think we need all this ceremony? Could we just use `rand.Intn(5)` below or `rand.Float64() * 5.0`?

Reply:
Some customers (eg. IBM who requested it)have very large numbers (1000s) and this cache was introduced due to thundering herd problems they'd have when refreshing things and all the PKI certs would get fetched at once. The cache helps but if they have all the same TTL it could still happen without jitter.There are several historical issues/fixes in consul-template for such cases and given the original problem I thought it would be a good addition.

Reply:
I mostly meant all of the ceremony around creating a `rand.New`, though I suppose if this ran very early all of the global `rand` would have the same seed and so would be likely to generate the same jitter... My inclination would be to do a `rand.Seed(time.Now().UnixNano())` call in an `init()` somewhere instead though.

Downgraded to a nit :)

Reply:
Yep. If you use rand.Intn() without seeding it, it would use the same value for all calls resulting in identical jitter values which defeats the purpose.